Generic Site settings Deployment Step

agriffard opened the pull request:
Fixes #5325, #5558

/cc  @sebastienros @deanmarcussen As explained during today's meeting.

agriffard commented:
Just implemented it in OrchardCore.Admin with AdminSetting for the moment.

Please tell me if it is the correct way to do it:

```
    [RequireFeatures("OrchardCore.Deployment")]
    public class DeploymentStartup : StartupBase
    {
        public override void ConfigureServices(IServiceCollection services)
        {
            var serviceProvider = services.BuildServiceProvider();
            var S = serviceProvider.GetService<IStringLocalizer<Startup>>();
            services.AddSingleton<IDeploymentStepFactory>(s => new GenericSiteSettingsDeploymentStepFactory(nameof(AdminSettings), S["Admin settings"], S["Exports the admin site settings."]));
        }
    }
```

agriffard commented:
Add Step:
![image](https://user-images.githubusercontent.com/703248/86407537-8550f800-bcb5-11ea-8c08-fab89b319fab.png)

Edit Step:
![image](https://user-images.githubusercontent.com/703248/86407439-56d31d00-bcb5-11ea-9105-de80c3942674.png)

Deployment plan:
![image](https://user-images.githubusercontent.com/703248/86407464-66526600-bcb5-11ea-93e0-37579685f459.png)

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Settings/Deployment/GenericSiteSettingsDeploymentStep.cs:
```diff
@@ -0,0 +1,25 @@
+    /// <summary>
+    /// Adds a generic site settings to a <see cref="DeploymentPlanResult"/>.
+    /// </summary>
+    public class GenericSiteSettingsDeploymentStep : DeploymentStep
```
```suggestion
    public class GenericSiteSettingsDeploymentStep<TModel> : DeploymentStep where TModel : class
```

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Admin/Startup.cs:
```diff
@@ -81,4 +85,15 @@ public override void ConfigureServices(IServiceCollection services)
             });
         }
     }
+
```
We're going to have to move al to a new project `OrchardCore.Settings.Core` because the modules shouldn't be referencing `OrchardCore.Settings`. Not sure why the `Admin` module has that reference already, we should fix that too :)

deanmarcussen commented:
```
        public override void ConfigureServices(IServiceCollection services)
        {
            services.AddSingleton<IDeploymentStepFactory>(sp => {
           var S = sp.GetService<IStringLocalizer<Startup>>();
 
return new GenericSiteSettingsDeploymentStepFactory(nameof(AdminSettings), S["Admin settings"], S["Exports the admin site settings."])
});
        }
```

For infos. But also will have to be generics, and I think in this case we can get away with out a custom factory, just a generic class, but would have to check

agriffard commented:
Added OrchardCore.Settings.Core with generic SiteSettingsDeploymentStep, source and driver.
Renamed the existing SiteSettingsDeploymentStep in OrchardCore.Settings into GeneralSiteSettingsDeploymentStep.

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Admin/Startup.cs:
```diff
@@ -81,4 +85,19 @@ public override void ConfigureServices(IServiceCollection services)
+            services.AddScoped<IDisplayDriver<DeploymentStep>>(sp =>
+            {
+                var S = sp.GetService<IStringLocalizer<Startup>>();
+                return new SiteSettingsPropertyDeploymentStepDriver<AdminSettings>(S["Admin settings"], S["Exports the admin site settings."]);
```
The text here should probably be
`Exports the admin settings`

Users, shouldn't need to know these are stored on the `site` property. That's why we're doing the seperate steps this way, is so that users/admins don't need to understand the distinction.

It's ok that developers need to understand

agriffard replied:
ok

agriffard commented on src/OrchardCore/OrchardCore.Settings.Core/Deployment/SiteSettingsPropertyDeploymentStepDriver.cs:
```diff
@@ -0,0 +1,46 @@
+        public override IDisplayResult Display(SiteSettingsPropertyDeploymentStep<TModel> step)
+        {
+            // TODO move this initialization to a BuildViewModel method.
+            return
```
Please do it.

deanmarcussen replied:
Done

agriffard commented on src/OrchardCore.Modules/OrchardCore.Users/Startup.cs:
```diff
@@ -311,6 +344,14 @@ public override void ConfigureServices(IServiceCollection services)
             services.AddScoped<INavigationProvider, ResetPasswordAdminMenu>();
             services.AddScoped<IDisplayDriver<ISite>, ResetPasswordSettingsDisplayDriver>();
+          //TODO this needs to be a seperate startup, just for the deployment step
+            services.AddTransient<IDeploymentSource, SiteSettingsPropertyDeploymentSource<ResetPasswordSettings>>();
```
ok

agriffard commented on src/OrchardCore.Modules/OrchardCore.Users/Startup.cs:
```diff
@@ -263,10 +287,19 @@ public override void ConfigureServices(IServiceCollection services)
             services.AddScoped<INavigationProvider, RegistrationAdminMenu>();
             services.AddScoped<IDisplayDriver<ISite>, RegistrationSettingsDisplayDriver>();
+          //TODO this needs to be a seperate startup, just for the deployment step
+            services.AddTransient<IDeploymentSource, SiteSettingsPropertyDeploymentSource<RegistrationSettings>>();
```
ok

agriffard commented on src/OrchardCore.Modules/OrchardCore.Users/Startup.cs:
```diff
@@ -236,10 +249,21 @@ public override void ConfigureServices(IServiceCollection services)
             services.AddScoped<IDisplayDriver<ISite>, ChangeEmailSettingsDisplayDriver>();
+            //TODO this needs to be a seperate startup, just for the deployment step
+            // [Feature("OrchardCore.Users.ChangeEmail")]
+            // [RequireFeatures("OrchardCore.Deployment")]
```
ok

agriffard commented on src/OrchardCore.Modules/OrchardCore.Users/Startup.cs:
```diff
@@ -189,6 +193,14 @@ public override void ConfigureServices(IServiceCollection services)
 
             services.AddScoped<IThemeSelector, UsersThemeSelector>();
+
+            services.AddTransient<IDeploymentSource, SiteSettingsPropertyDeploymentSource<LoginSettings>>();
```
ok, please check.

agriffard commented on src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Startup.cs:
```diff
@@ -66,5 +70,19 @@ public override void Configure(IApplicationBuilder builder, IEndpointRouteBuilde
+        {
+            public override void ConfigureServices(IServiceCollection services)
+            {
+                services.AddTransient<IDeploymentSource, SiteSettingsPropertyDeploymentSource<BackgroundTaskSettings>>();
```
ok

agriffard commented on src/OrchardCore.Modules/OrchardCore.Users/Startup.cs:
```diff
@@ -38,6 +41,7 @@
 
 namespace OrchardCore.Users
 {
+    // [RequireFeatures("OrchardCore.Deployment")]
```
ok

agriffard commented:
![image](https://user-images.githubusercontent.com/703248/86511523-5ee8a500-bdfa-11ea-8ff7-3b221dc036d6.png)

deanmarcussen commented:
Merged dev, and added the _Taxonomy Filters settings_ too

agriffard commented:
Do we keep the GraphQL settings?

deanmarcussen commented:
@agriffard I have evaluated and updated your list on https://github.com/OrchardCMS/OrchardCore/issues/5558

Some of the ones you have done here, have their own recipe step, so cannot be included in a generic design.

Others have secrets, so also must be custom.

GraphQL is not a Site Settings storage, it's configured via `IOptions` so was just not named as per the convention originally.

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/OrchardCore.Apis.GraphQL.csproj:
```diff
@@ -19,6 +19,7 @@
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.DisplayManagement\OrchardCore.DisplayManagement.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Module.Targets\OrchardCore.Module.Targets.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Navigation.Core\OrchardCore.Navigation.Core.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Settings.Core\OrchardCore.Settings.Core.csproj" />
```
Remove this reference

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.BackgroundTasks/OrchardCore.BackgroundTasks.csproj:
```diff
@@ -16,6 +16,7 @@
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.DisplayManagement\OrchardCore.DisplayManagement.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Navigation.Core\OrchardCore.Navigation.Core.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.ResourceManagement\OrchardCore.ResourceManagement.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Settings.Core\OrchardCore.Settings.Core.csproj" />
```
Remove this reference

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Facebook/OrchardCore.Facebook.csproj:
```diff
@@ -21,6 +21,7 @@
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Navigation.Core\OrchardCore.Navigation.Core.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.ResourceManagement\OrchardCore.ResourceManagement.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Recipes.Abstractions\OrchardCore.Recipes.Abstractions.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Settings.Core\OrchardCore.Settings.Core.csproj" />
```
same

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.GitHub/OrchardCore.GitHub.csproj:
```diff
@@ -18,6 +18,7 @@
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Navigation.Core\OrchardCore.Navigation.Core.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.ResourceManagement\OrchardCore.ResourceManagement.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Recipes.Abstractions\OrchardCore.Recipes.Abstractions.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Settings.Core\OrchardCore.Settings.Core.csproj" />
```
same

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Google/OrchardCore.Google.csproj:
```diff
@@ -17,6 +17,7 @@
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Navigation.Core\OrchardCore.Navigation.Core.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.ResourceManagement\OrchardCore.ResourceManagement.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Recipes.Abstractions\OrchardCore.Recipes.Abstractions.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Settings.Core\OrchardCore.Settings.Core.csproj" />
```
same

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Microsoft.Authentication/OrchardCore.Microsoft.Authentication.csproj:
```diff
@@ -18,6 +18,7 @@
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Navigation.Core\OrchardCore.Navigation.Core.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.ResourceManagement\OrchardCore.ResourceManagement.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Recipes.Abstractions\OrchardCore.Recipes.Abstractions.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Settings.Core\OrchardCore.Settings.Core.csproj" />
```
same

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Twitter/OrchardCore.Twitter.csproj:
```diff
@@ -18,6 +18,7 @@
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Navigation.Core\OrchardCore.Navigation.Core.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.ResourceManagement\OrchardCore.ResourceManagement.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Recipes.Abstractions\OrchardCore.Recipes.Abstractions.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Settings.Core\OrchardCore.Settings.Core.csproj" />
```
same

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Settings/Views/Items/SiteSettingsPropertyDeploymentStep.Fields.Edit.cshtml:
```diff
@@ -0,0 +1,6 @@
+<h5>@Model.Title</h5>
+
+<div class="form-group">
+    <span class="hint">@Model.Description</span>
```
Tested is all good

deanmarcussen commented on src/OrchardCore/OrchardCore.Settings.Core/Deployment/SiteSettingsPropertyDeploymentSource.cs:
```diff
@@ -0,0 +1,33 @@
+
+            var siteSettings = await _siteService.GetSiteSettingsAsync();
+
+            result.Steps.Add(new JObject(
```
We should find the property if it exists already, so it is only one step that is being added.
I will do this, while I am checking the registration with both attributes (i.e. `Feature` and `RequiresFeature` works ok)

deanmarcussen replied:
@agriffard have changed it to look for an existings `Settings` property rather than creating a new one every time.

output changes from
```
  "steps": [
    {
      "name": "Settings", <-- first Settings step
      "AdminSettings": {
        "DisplayMenuFilter": false
      }
    },
    {
      "name": "Settings", <-- second Settings step
      "LoginSettings": {
        "UseSiteTheme": false,
        "UseExternalProviderIfOnlyOneDefined": false,
        "DisableLocalLogin": false,
        "UseScriptToSyncRoles": false,
        "SyncRolesScript": null
      }
    }
  ]
```

```
  "steps": [
    {
      "name": "Settings", <-- Single Settings step
      "AdminSettings": {
        "DisplayMenuFilter": false
      },
      "LoginSettings": {
        "UseSiteTheme": false,
        "UseExternalProviderIfOnlyOneDefined": false,
        "DisableLocalLogin": false,
        "UseScriptToSyncRoles": false,
        "SyncRolesScript": null
      },
      "HttpsSettings": {
        "EnableStrictTransportSecurity": false,
        "RequireHttps": false,
        "RequireHttpsPermanent": false,
        "SslPort": null
      }
    }
  ]
```

An optimization, and makes it clearer that all the steps are contained in one 

If you are happy with the change let me know, and I think we can merge it.

I've checked the `Feature` and `RequiresFeature` are working fine together. i.e. if you disable deployment but still have registration the startup doesn't fire.

agriffard replied:
Great

agriffard commented:
I prefixed the Factory key with the name instead of suffixing it :
`private static readonly string GenericTypeKey = typeof(TModel).Name + "_SiteSettingsPropertyDeploymentStep";`

Thanks to this, the steps will be sorted a bit more alphabetically in the modal.

deanmarcussen commented:
> I prefixed the Factory key with the name instead of suffixing it
> Thanks to this, the steps will be sorted a bit more alphabetically in the modal.

Cool, can you do the same in the
```
        public SiteSettingsPropertyDeploymentStep()
        {
            Name = "SiteSettingsPropertyDeploymentStep_" + typeof(TModel).Name;
        }
```

so the match I have a vague memory this is important ;)
